scripts/cot/botania/flowers/genereting/nuclianthus.zs

```diff
@@ -0,0 +1,138 @@
+/*
+Nuclianthus (Helianthus + nuclear) - botanical reactor
+Generates power using nuclear fuel (produces nuclear waste)
+Catch: if buffer is overflowing (for some mana amount wasted - but heals overtime) it explodes KABOOM!!!
+
+Stats:
+Mana per 1 sec: H/t*eff
+Time working: as on fuel
+*/
+#loader contenttweaker
+
+import crafttweaker.data.IData;
+import crafttweaker.entity.IEntityItem;
+import crafttweaker.item.IItemStack;
+import crafttweaker.util.Math;
+import crafttweaker.world.IBlockPos;
+import crafttweaker.world.IWorld;
+import mods.contenttweaker.VanillaFactory;
+import mods.randomtweaker.cote.SubTileEntityInGame;
+import mods.zenutils.DataUpdateOperation.OVERWRITE;
+import native.nc.enumm.IFissionFuelEnum;
+import native.nc.item.ItemFissionFuel;
+import native.net.minecraft.util.EnumParticleTypes;
+import native.net.minecraft.world.WorldServer;
+
+static fuelsList as int[][string] = {
+/*
+Fromat:
+fuel:name [duration, heat/t, effciency]
+*/
+
+  // IC2 fuels
+  'ic2.nuclear.uranium': [7600,  100, 120],    // Enriched uranium
+  'ic2.nuclear.mox'    : [45600, 140, 135],    // MOX
+
+} as int[][string];
+
+static fuelDurationMultiplier as double = 1.0;
+static fuelManaGenerationMultiplier as double = 1.0;
+static overHeatLimit as int = 10000;
+
+val nuclianthus = VanillaFactory.createSubTileGenerating('nuclianthus', 0xFFFF00);
+nuclianthus.maxMana = 10000;
+nuclianthus.passiveFlower = false;
+nuclianthus.range = 1;
+nuclianthus.onUpdate = function (subtile, world, pos) {
+  if (world.isRemote()) return;
+  isWorking(subtile) ? generate(world, pos, subtile) : pickUpFuel(world, pos, subtile);
+};
+nuclianthus.register();
+
+function isWorking(subtile as SubTileEntityInGame) as bool {
+  return !(isNull(subtile.data)
+    || isNull(subtile.data.Status)
+    || subtile.data.Status != 'work');
+}
+
+function generate(world as IWorld, pos as IBlockPos, subtile as SubTileEntityInGame) as void {
+  if (subtile.data.FuelData.duration > 0) {
+    val manaGenerated = (subtile.data.FuelData.production / 20) as int;
+    val heatGenerated = Math.max(manaGenerated + subtile.getMana() - subtile.getMaxMana(), 0);
+    subtile.setCustomData(subtile.data.deepUpdate({Overheat: (Math.max((subtile.data.Overheat + heatGenerated - 10) as int , 0)), FuelData: {duration: subtile.data.FuelData.duration - 2}},{FuelData: {duration: OVERWRITE}, Overheat: OVERWRITE}));
+    subtile.addMana(manaGenerated);
+    if (world.random.nextInt(5) > 2) scripts.lib.sound.play('minecraft:block.lava.pop', pos, world); // I would like to use this: "nuclearcraft:player.geiger_tick"
+    (world.native as WorldServer).spawnParticle(EnumParticleTypes.REDSTONE, 0.5f + pos.x, 1.2 + pos.y, 0.5f + pos.z, 0, ((subtile.data.Overheat as float) / 10000) - 1, (((10000 - subtile.data.Overheat) as float) / 10000), 0, 1, 0);
```

Comment:
Particles generated in wrong place - slightly shifted from center of the flower:
![](https://i.imgur.com/qoikjdv.png)

My suggestion for better particles: make several particle "sources" around the flower, depending on amount of time left.

1. For example, when you just dropped weak fuel, it generates 1 source of particle "clound" - on center over the flower
![](https://i.imgur.com/s53CQnP.png)

2. Then, if fuel is last longer, there is 2 particle clouds
![](https://i.imgur.com/i17gF0b.png)

3. Amount of clouds could be square rooted of time. For example
  - 1 cloud: 2 seconds left
  - 2 clouds: 4 seconds left
  - 3 clouds: 8 seconds left

Advantages:
- Clearly see if fuel is better than another tested one
- Clearly see how much time left